File: packages/cli/src/templates/cc-native/_cc-native/lib/reviewers/agent.py
"""
CC-Native Agent Reviewer Module.

Runs Claude Code agents to review plans.
"""

import json
import shutil
import subprocess
import sys
from pathlib import Path
from typing import Any, Dict, Optional

# Import from parent lib
_lib_dir = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(_lib_dir))

from utils import ReviewerResult, parse_json_maybe, coerce_to_review
from debug import debug_log, debug_raw
from .base import AgentConfig, AGENT_REVIEW_PROMPT_PREFIX

# Import logger
_shared_logger = Path(__file__).resolve().parent.parent.parent.parent / "_shared" / "lib"
sys.path.insert(0, str(_shared_logger))
from base.logger import log_debug, log_info, log_warn, log_error

# Import shared subprocess utilities
_shared_lib = Path(__file__).resolve().parent.parent.parent.parent / "_shared" / "lib" / "base"
sys.path.insert(0, str(_shared_lib))
from subprocess_utils import get_internal_subprocess_env


def _parse_claude_output(raw: str) -> Optional[Dict[str, Any]]:
    """Parse Claude CLI JSON output, handling various formats.

    Claude CLI can output in several formats:
    - Direct structured_output dict
    - Assistant message with StructuredOutput tool use
    - List of events with assistant messages

    Args:
        raw: Raw stdout from Claude CLI

    Returns:
        Parsed JSON dict or None if parsing failed
    """
    try:
        result = json.loads(raw)
        if isinstance(result, dict):
            if "structured_output" in result:
                log_debug("agent", "Found structured_output in root dict", component="parse")
                return result["structured_output"]
            if result.get("type") == "assistant":
                message = result.get("message", {})
                content = message.get("content", [])
                for item in content:
                    if isinstance(item, dict) and item.get("name") == "StructuredOutput":
                        log_debug("agent", "Found StructuredOutput in assistant message content", component="parse")
                        return item.get("input", {})
                log_debug("agent", "Assistant message found but no StructuredOutput tool use in content", component="parse")
        elif isinstance(result, list):
            log_debug("agent", f"Received list of {len(result)} events, searching for assistant message", component="parse")
            for i, event in enumerate(result):
                if not isinstance(event, dict):
                    continue
                if event.get("type") == "assistant":
                    message = event.get("message", {})
                    content = message.get("content", [])
                    for item in content:
                        if isinstance(item, dict) and item.get("name") == "StructuredOutput":
                            log_debug("agent", f"Found StructuredOutput in event[{i}] assistant message", component="parse")
                            return item.get("input", {})
            log_debug("agent", "No StructuredOutput found in any assistant message in event list", component="parse")
    except json.JSONDecodeError as e:
        log_warn("agent", f"JSON decode error: {e}", component="parse")
    except Exception as e:
        log_error("agent", f"Unexpected error during structured parsing: {e}", component="parse")

    # Fallback to heuristic extraction with required field validation
    log_debug("agent", "No structured output found, falling back to heuristic JSON extraction", component="parse")
    return parse_json_maybe(raw, require_fields=["verdict", "summary"])


def run_agent_review(
    plan: str,
    agent: AgentConfig,
    schema: Dict[str, Any],
    timeout: int,
    context_path: Optional[Path] = None,
    session_name: str = "unknown",
) -> ReviewerResult:
    """Run a single Claude Code agent to review the plan.

    Args:
        plan: The plan content to review
        agent: Agent configuration (name, model, etc.)
        schema: JSON schema for the review output
        timeout: Timeout in seconds
        context_path: Optional path to context folder for debug logging
        session_name: Session name for debug logging

    Returns:
        ReviewerResult with the review output
    """
    claude_path = shutil.which("claude")
    if claude_path is None:
        log_warn(agent.name, "Claude CLI not found on PATH")
        return ReviewerResult(
            name=agent.name,
            ok=False,
            verdict="skip",
            data={},
            raw="",
            err="claude CLI not found on PATH",
        )

    log_debug(agent.name, f"Found Claude CLI at: {claude_path}")

    # User prompt - direct instruction to call StructuredOutput immediately
    prompt = f"""IMMEDIATELY call StructuredOutput with your review of the plan below.
Do NOT output any text before calling StructuredOutput.

PLAN:
<<<
{plan}
>>>
"""

    schema_json = json.dumps(schema, ensure_ascii=False)

    # Build command args - use --system-prompt with the markdown body as persona
    cmd_args = [
        claude_path,
        "-p",  # Enable print mode to read prompt from stdin
        "--model", agent.model,
        "--output-format", "json",
        "--json-schema", schema_json,
        "--max-turns", "3",  # Allow buffer for tool call + result (usually completes in 2)
        "--setting-sources", "",  # Disable user/project settings to avoid PAI context interference
    ]

    # Add system prompt: prefix with single-turn instructions, then agent's persona
    if agent.system_prompt:
        full_prompt = AGENT_REVIEW_PROMPT_PREFIX + "\n\n---\n\n" + agent.system_prompt
        cmd_args.extend(["--system-prompt", full_prompt])

    log_info(agent.name, f"Running with model: {agent.model}, timeout: {timeout}s")

    # Get environment for internal subprocess (bypasses hooks)
    env = get_internal_subprocess_env()

    try:
        p = subprocess.run(
            cmd_args,
            input=prompt,
            text=True,
            capture_output=True,
            timeout=timeout,
            encoding="utf-8",
            errors="replace",
            env=env,
        )
    except subprocess.TimeoutExpired:
        log_warn(agent.name, f"TIMEOUT after {timeout}s")
        return ReviewerResult(agent.name, False, "error", {}, "", f"{agent.name} timed out after {timeout}s")
    except Exception as ex:
        log_error(agent.name, f"Exception: {ex}")
        return ReviewerResult(agent.name, False, "error", {}, "", f"{agent.name} failed to run: {ex}")

    log_debug(agent.name, f"Exit code: {p.returncode}")
    log_debug(agent.name, f"stdout length: {len(p.stdout or '')} chars")
    if p.stderr:
        log_debug(agent.name, f"stderr: {p.stderr[:500]}")

    raw = (p.stdout or "").strip()
    err = (p.stderr or "").strip()

    # Debug logging - capture full raw output for diagnosis
    if context_path:
        debug_raw(context_path, session_name, f"agent:{agent.name}", "stdout", raw)
        if err:
            debug_raw(context_path, session_name, f"agent:{agent.name}", "stderr", err)
        debug_log(context_path, session_name, f"agent:{agent.name}", "subprocess_info", {
            "exit_code": p.returncode,
            "stdout_len": len(raw),
            "stderr_len": len(err),
            "model": agent.model,
            "timeout": timeout,
        })

    if raw:
        log_debug(agent.name, f"stdout preview: {raw[:500]}")

    obj = _parse_claude_output(raw)

    # Debug logging - capture parsed result details
    if context_path:
        debug_log(context_path, session_name, f"agent:{agent.name}", "parsed_result", {
            "parsed_keys": list(obj.keys()) if obj else None,
            "verdict": obj.get("verdict") if obj else None,
            "has_summary": bool(obj.get("summary")) if obj else False,
            "summary_preview": (obj.get("summary", "")[:200] + "...") if obj and obj.get("summary") and len(obj.get("summary", "")) > 200 else (obj.get("summary") if obj else None),
            "issues_count": len(obj.get("issues", [])) if obj else 0,
            "missing_sections_count": len(obj.get("missing_sections", [])) if obj else 0,
            "questions_count": len(obj.get("questions", [])) if obj else 0,
        })

    if obj:
        log_info(agent.name, f"Parsed JSON successfully, verdict: {obj.get('verdict', 'N/A')}")
    else:
        log_warn(agent.name, "Failed to parse JSON from output")

    ok, verdict, norm = coerce_to_review(obj, "Retry or check agent configuration.")

    return ReviewerResult(agent.name, ok, verdict, norm, raw, err)

File: packages/cli/src/templates/cc-native/_cc-native/lib-ts/reviewers/agent.ts
/**
 * Agent-based plan reviewer with multi-provider support.
 * Routes to Claude or Codex CLI based on agent.provider field.
 * See cc-native-plan-review-spec.md §4.10
 */

import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import { logDebug, logInfo, logWarn, logError } from "../../../_shared/lib-ts/base/logger.js";
import { getInternalSubprocessEnv, findExecutable, execFileAsync, shellQuoteWin } from "../../../_shared/lib-ts/base/subprocess-utils.js";
import { parseCliOutput } from "../cli-output-parser.js";
import { parseJsonMaybe, coerceToReview } from "../json-parser.js";
import { debugLog, debugRaw } from "../debug.js";
import type { AgentConfig, ReviewerResult, ReviewOptions } from "../types.js";
import { AGENT_REVIEW_PROMPT_PREFIX } from "../types.js";
import { makeResult } from "./types.js";
import type { Reviewer } from "./types.js";

/**
 * Agent reviewer — runs a CLI instance with a custom persona.
 */
export class AgentReviewer implements Reviewer {
  constructor(private agent: AgentConfig) {}

  async review(
    plan: string,
    schema: Record<string, unknown>,
    options: ReviewOptions,
  ): Promise<ReviewerResult> {
    return runAgentReview(
      plan,
      this.agent,
      schema,
      options.timeout,
      options.context_path,
      options.session_name ?? "unknown",
    );
  }
}

/**
 * Run a single agent to review the plan.
 * Routes to provider-specific implementation based on agent.provider.
 * Never throws — returns error ReviewerResult on failure.
 */
export async function runAgentReview(
  plan: string,
  agent: AgentConfig,
  schema: Record<string, unknown>,
  timeout: number,
  contextPath?: string,
  sessionName = "unknown",
): Promise<ReviewerResult> {
  if (agent.provider === "codex") {
    return runAgentReviewCodex(plan, agent, schema, timeout, contextPath, sessionName);
  }
  // Default: Claude (existing implementation)
  return runAgentReviewClaude(plan, agent, schema, timeout, contextPath, sessionName);
}

/**
 * Run a single Claude Code agent to review the plan.
 */
async function runAgentReviewClaude(
  plan: string,
  agent: AgentConfig,
  schema: Record<string, unknown>,
  timeout: number,
  contextPath?: string,
  sessionName = "unknown",
): Promise<ReviewerResult> {
  const claudePath = findExecutable("claude");
  if (!claudePath) {
    logWarn(agent.name, "Claude CLI not found on PATH");
    return makeResult(agent.name, false, "skip", {}, "", "claude CLI not found on PATH");
  }

  logDebug(agent.name, `Found Claude CLI at: ${claudePath}`);

  const prompt = `IMMEDIATELY call StructuredOutput with your review of the plan below.
Do NOT output any text before calling StructuredOutput.

PLAN:
<<<
${plan}
>>>
`;

  const schemaJson = JSON.stringify(schema);
  const cmdArgs = [
    "--model", agent.model,
    "--output-format", "json",
    "--json-schema", shellQuoteWin(schemaJson),
    "--max-turns", "3",
    "--setting-sources", process.platform === "win32" ? '""' : "",
    "-p",
  ];

  if (agent.system_prompt) {
    const fullPrompt = AGENT_REVIEW_PROMPT_PREFIX + "\n\n---\n\n" + agent.system_prompt;
    cmdArgs.push("--system-prompt", shellQuoteWin(fullPrompt));
  }

  logInfo(agent.name, `Running Claude with model: ${agent.model}, timeout: ${timeout}s`);

  const env = getInternalSubprocessEnv();

  const result = await execFileAsync(claudePath, cmdArgs, {
    input: prompt,
    timeout: timeout * 1000,
    env: env as Record<string, string>,
    maxBuffer: 10 * 1024 * 1024,
    shell: process.platform === "win32",
  });

  if (result.killed || result.signal === "SIGTERM") {
    logWarn(agent.name, `Claude TIMEOUT after ${timeout}s`);
    return makeResult(agent.name, false, "error", {}, "", `${agent.name} timed out after ${timeout}s`);
  }

  const raw = result.stdout.trim();
  const err = result.stderr.trim();

  if (!raw && !err && result.exitCode !== 0) {
    logError(agent.name, `Process exited with code ${result.exitCode} and no output`);
    return makeResult(agent.name, false, "error", {}, "", `${agent.name} failed to run (exit ${result.exitCode})`);
  }

  logDebug(agent.name, `Exit code: ${result.exitCode}`);
  logDebug(agent.name, `stdout length: ${raw.length} chars`);
  if (err) logDebug(agent.name, `stderr: ${err.slice(0, 500)}`);

  // Debug logging
  if (contextPath) {
    debugRaw(contextPath, sessionName, `agent:${agent.name}`, "stdout", raw);
    if (err) {
      debugRaw(contextPath, sessionName, `agent:${agent.name}`, "stderr", err);
    }
    debugLog(contextPath, sessionName, `agent:${agent.name}`, "subprocess_info", {
      exit_code: result.exitCode,
      stdout_len: raw.length,
      stderr_len: err.length,
      model: agent.model,
      provider: "claude",
      timeout,
    });
  }

  if (raw) logDebug(agent.name, `stdout preview: ${raw.slice(0, 500)}`);

  const obj = parseCliOutput(raw, ["verdict", "summary"]);

  if (contextPath) {
    debugLog(contextPath, sessionName, `agent:${agent.name}`, "parsed_result", {
      parsed_keys: obj ? Object.keys(obj) : null,
      verdict: obj?.verdict ?? null,
      has_summary: obj ? Boolean(obj.summary) : false,
      issues_count: obj && Array.isArray(obj.issues) ? (obj.issues as unknown[]).length : 0,
    });
  }

  if (obj) {
    logInfo(agent.name, `Parsed JSON successfully, verdict: ${obj.verdict ?? "N/A"}`);
  } else {
    logWarn(agent.name, "Failed to parse JSON from output");
  }

  const [ok, verdict, norm] = coerceToReview(
    obj as Record<string, unknown> | null,
    "Retry or check agent configuration.",
  );

  return makeResult(agent.name, ok, verdict, norm, raw, err);
}

/**
 * Run a single Codex CLI agent to review the plan.
 * Adapts the codex.ts reviewer pattern for agent-based review with persona.
 */
async function runAgentReviewCodex(
  plan: string,
  agent: AgentConfig,
  schema: Record<string, unknown>,
  timeout: number,
  contextPath?: string,
  sessionName = "unknown",
): Promise<ReviewerResult> {
  const codexPath = findExecutable("codex");
  if (!codexPath) {
    logWarn(agent.name, "Codex CLI not found on PATH, skipping");
    return makeResult(agent.name, false, "skip", {}, "", "codex CLI not found on PATH");
  }

  // Codex has no --system-prompt flag, so we prepend the agent persona to stdin.
  const fullPrompt = [
    AGENT_REVIEW_PROMPT_PREFIX,
    "---",
    agent.system_prompt || "",
    "---",
    `Return ONLY a JSON object matching this schema:\n${JSON.stringify(schema)}`,
    "",
    "PLAN:",
    "<<<",
    plan,
    ">>>",
  ].join("\n\n");

  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), `codex-agent-${agent.name}-`));

  try {
    const schemaPath = path.join(tmpDir, "schema.json");
    const outPath = path.join(tmpDir, "output.json");
    fs.writeFileSync(schemaPath, JSON.stringify(schema, null, 2), "utf-8");

    const cmdArgs = ["exec", "--sandbox", "read-only"];
    if (agent.model) cmdArgs.push("--model", agent.model);
    cmdArgs.push("--output-schema", schemaPath, "-o", outPath, "-");

    logInfo(agent.name, `Running Codex with model: ${agent.model}, timeout: ${timeout}s`);

    const env = getInternalSubprocessEnv();

    const result = await execFileAsync(codexPath, cmdArgs, {
      input: fullPrompt,
      timeout: timeout * 1000,
      env: env as Record<string, string>,
      maxBuffer: 10 * 1024 * 1024,
      shell: process.platform === "win32",
    });

    if (result.killed || result.signal === "SIGTERM") {
      logWarn(agent.name, `Codex TIMEOUT after ${timeout}s`);
      return makeResult(agent.name, false, "error", {}, "", `${agent.name} (codex) timed out after ${timeout}s`);
    }

    const err = result.stderr.trim();

    // Log exit code and stderr tail for ALL non-zero exits (aids diagnosis of intermittent failures)
    if (result.exitCode !== 0) {
      const stderrTail = err.slice(-500);
      logWarn(agent.name, `Codex exited with code ${result.exitCode}, stderr_len=${err.length}, stderr_tail: ${stderrTail}`);
    }

    if (!result.stdout && !err && !fs.existsSync(outPath) && result.exitCode !== 0) {
      logError(agent.name, `Codex exited with code ${result.exitCode} and no output`);
      return makeResult(agent.name, false, "error", {}, "", `${agent.name} (codex) failed (exit ${result.exitCode})`);
    }

    // Read output: prefer temp file, fallback to stdout
    let raw = "";
    const outExists = fs.existsSync(outPath);
    if (outExists) {
      raw = fs.readFileSync(outPath, "utf-8");
    }

    logDebug(agent.name, `Codex output: exit=${result.exitCode}, outFile=${outExists} (${raw.length} chars), stdout=${result.stdout.length} chars`);

    // Debug logging
    if (contextPath) {
      debugRaw(contextPath, sessionName, `agent:${agent.name}`, "stdout", raw || result.stdout);
      if (err) debugRaw(contextPath, sessionName, `agent:${agent.name}`, "stderr", err);
      debugLog(contextPath, sessionName, `agent:${agent.name}`, "subprocess_info", {
        exit_code: result.exitCode,
        stdout_len: (raw || result.stdout).length,
        stderr_len: err.length,
        out_file_exists: outExists,
        model: agent.model,
        provider: "codex",
        timeout,
      });
    }

    // Parse output
    const obj = parseJsonMaybe(raw) ?? parseJsonMaybe(result.stdout);
    const [ok, verdict, norm] = coerceToReview(
      obj,
      "Retry or check Codex CLI auth/config.",
    );

    return makeResult(agent.name, ok, verdict, norm, raw || result.stdout, err);
  } finally {
    try {
      fs.rmSync(tmpDir, { recursive: true, force: true });
    } catch (e) {
      logDebug(agent.name, `Failed to cleanup temp dir ${tmpDir}: ${e}`);
    }
  }
}
